First Commit

Harshmanitripathi opened the pull request:
Description of work done on Assignment 2 student dataBase
1. First In-order to create ArrayOfObjects I creating a class of name(details) and specified it with different characteristics 
     of student such as rollno. , name, age , address , and courses(A,B,C,D,E,F).
2 . I used scanner class to take the input and for assigning of courses as the courses assigned to each student are different 
      so, I assigned default value as false and whatever the user select as course I assigned that course true for the particular    
      user, the input of the user for the courses I took in the queue and for showcasing the courses left I uses set 
      dataStructure.
3. By creating the class of name "details" I used it to create array of objects in which all the details of the student are 
   assigned in form of object at particular index and all the students details can be accessed from there.
4. For sorting from different categories like age , rollno, name , etc .I used **Arrays.sort()** method implements 
  **Comparator<>** interface and used the abstract method **compareTo()** to compare the two different parameter of 
   object and sort them completely.
5. Furthur I used serialization interface to store the data in txt file and retrieve it but only problem faced was as the data was needed to be stored in object and as I created the object it is refresing the new data with previous one . 
    whole content of object inside array.

Commit:
First Commit

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
-    public static void main(String[] args) {
-        System.out.println("Hello world!");
+    public static void main(String[] args) throws IOException, ClassNotFoundException {
+
```
First point. Do not code everything in one class
Its a project and not a program.

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+        int noOfStudents = sc.nextInt();
+        int rollNo=0,ageOfStudent=0;
+        String addressOfStudent="",nameOfStudent="";
+        boolean courseA = false,courseB= false,courseC = false,courseD=false,courseE = false,courseF = false;
```
Utilize OOPS
Create a Student class and use an array/list of objects here based on the number of students input 
Courses should be an array or set of characters

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+            if(noOfCourses<4){
+                System.out.println("Please enter at least 4 courses");
+                while(noOfCourses<4)
+                    noOfCourses = sc.nextInt();
```
Read about single responsibility clause
Classes and functions need to have purposes
Keep all input and output as part of one class

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+        System.out.println();
+    }
+
+}
```
Output should be part of a different class
Create a class for input and output.
Create a utility class for displaying borders, tables, formats etc

ajayvenkat10 replied:
Why is this function in this class 
Displaying a list of objects should not be part of the inner object class 
It should be responsible for only its own details and not other objects of the same type 
You can call this function for each object in the list of objects 
Also check the access specifier you give for methods and member variables 


If its used only in that class or by an object of the class, then you have to make it private 
Else based on use case it needs to be public or protected 

Read about access specifiers in Java

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+        this.courseD = courseD;
+        this.courseE = courseE;
+        this.courseF = courseF;
+    }
```
Why a separate public method? 
Use constructors for setting member all member variables for an instance of an object 
Again Courses should be an array/set and don't list them as separate. 
Imagine I ask you to keep increasing the number of subjects ? Consider scaling 
You are creating an application and its collaborative, not a lab program anymore
Think of it as a product.

Your code should be easy to read, reusable, scalable, risk free and bug free. Should be easy to modify and scale

Name the class right 
Also don't keep multiple classes in one file 
Each class should be a different file ideally

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+    public int compare(details o1, details o2) {
+        return o1.addressOfStudent.compareTo(o2.addressOfStudent);
+    }
+}
```
Again should be part of different classes

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+                }
+                else if(course.peek().equals("F")){
+                    courseF = true;
+                    course.remove();
```
Input should be part of a different class

ajayvenkat10 commented on src/Main.java:
```diff
@@ -1,5 +1,252 @@
+//        obj.courseF = courseF;
+//Saving the object in file
+
+        File F = new File("obj.txt");
+        FileOutputStream fileOut = new FileOutputStream(F);
+        ObjectOutputStream out = new ObjectOutputStream(fileOut);
+        out.writeObject(student);
+
+//        Retrive the object from file
+        FileInputStream fis = new FileInputStream(F);
+        ObjectInputStream ois = new ObjectInputStream(fis);
+        details obj1 = (details) ois.readObject();
+//        int[] showingDataStored = (int[]) ois.readObject();
+        ois.close();
+        fis.close();
+        System.out.println("Showing data which was saved in txt file");
+//        System.out.println(obj1.rollNoOfStudent);
```
Move it to a different File IO class

ajayvenkat10 reviewed (changes requested):
Added comments.

Commit:
Second Commit

Harshmanitripathi commented:
New features added:
1. Made the user to delete as many student he can.
2. Made user to have as many courses as he wants.
3. Saved the data using Serialization interface, but it is not able to store in form of data in form of array of object.

Changes made:
1.  Firstly the main class has only one main function and it call all the other functions required.

2. I have changed the name of all the class to start with capital letter.

3. I have created seperate class of "Student" to set the details of the particular student entered by user.

4. Created seperate classes for Sorting operation.

5. Deleting function is working well .

ajayvenkat10 commented on src/StudentPackage/Student.java:
```diff
@@ -0,0 +1,70 @@
+
+
+
+    public void setDetails(int rollNoOfStudent, int ageOfStudent, String nameOfStudent, String addressOfStudent, Queue<String> course)
+    {
+        this.rollNoOfStudent = rollNoOfStudent;
+        this.ageOfStudent = ageOfStudent;
+        this.nameOfStudent = nameOfStudent;
+        this.addressOfStudent = addressOfStudent;
+        int i=0;
+//        converting set to Array
+        while(course.size() > 0){
+            this.coursesTakenByUser[i++] = course.peek();
+            course.remove();
+
+        }
```
Use constructor or getter and setter 
Don't use another method to set details

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/Main.java:
```diff
@@ -0,0 +1,18 @@
+    public static void main(String[] args) throws IOException {
+
+//        Giving Input from User
+        TakingInputFromUser.TakingInput();
```
Rename `TakingInputFromUser` class

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/DisplayingOutputFromUser.java:
```diff
@@ -0,0 +1,14 @@
+            System.out.print(courses+", ");
+        }
+        System.out.println();
+    }
```
Don't need a separate class for output 
You can use one class for input and output 
Also you can have a display function inside the student class which will display the details of the student 
You can loop through array of students and call `student[i].displayDetails()`.

ajayvenkat10 commented on src/StudentPackage/DeletingStudents.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.*;
+
+public class DeletingStudents {
+
+
+    public static void studentDeleted(Student[] student, int rollNoToBeDeleted, boolean checkForRollNo) {
+
+        while (checkForRollNo == false){
+
+            for(int i=0;i<student.length;i++){
+                if(student[i] == null)break;
+                if (student[i].getRollNoOfStudent() == rollNoToBeDeleted){
+                    checkForRollNo = true;
+
+                    for (int j=i;j<student.length-1;j++){
+
+                        student[j] = student[j+1];
+                    }
+                    student[student.length-1] = null;
+                    break;
+                }
+            }
+            if (checkForRollNo == false)
+            {
+
+                System.out.println("Please enter the valid rollNo");
+                Scanner sc = new Scanner(System.in);
+                rollNoToBeDeleted = sc.nextInt();
+            }
+        }
+
+    }
+}
```
Don't make each operation a class 
Operation is a function 
You can have all these functions in an operations class 
Add, Delete, Display etc

ajayvenkat10 commented on src/StudentPackage/TakingInputFromUser.java:
```diff
@@ -0,0 +1,111 @@
+import java.io.IOException;
+import java.util.*;
+import java.util.Comparator;
+public class TakingInputFromUser {
```
Rename class 
Don't add verbs like Taking
Call it something like UserIO or InputOutput

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/TakingInputFromUser.java:
```diff
@@ -0,0 +1,111 @@
+
+            while (true) {
+                String inputFromUserAboutTheCourses = sc.next();
+                if (((inputFromUserAboutTheCourses).toLowerCase()).equals("done"))
```
Better to use a single character like (Y/N) instead of the word "done" 
You are building a user facing application, whenever you are expecting a user to type something in for your validation always keep it simple. Having type in a word can be more error prone than having to type in a letter

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/TakingInputFromUser.java:
```diff
@@ -0,0 +1,111 @@
+
+            }
+
+            student[i].setDetails(rollNo, ageOfStudent, nameOfStudent, addressOfStudent, setOfCourses);
```
Use constructors or use getters and setters. Don't use another method to initialize the object with details

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/TakingInputFromUser.java:
```diff
@@ -0,0 +1,111 @@
+        int noOfStudents = sc.nextInt();
+        int rollNo = 0, ageOfStudent = 0;
+        String addressOfStudent = "", nameOfStudent = "";
+        Queue<String> setOfCourses = new LinkedList<>();
```
Why do you need a Queue or a LinkedList ?

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/TakingInputFromUser.java:
```diff
@@ -0,0 +1,111 @@
+
+        System.out.println("enter the way you want to sort by age=1, name=2, rollno=3, address=4 etc.");
+        int sortBy = sc.nextInt();
+        switch (sortBy) {
+            case 1:
+                Arrays.sort(student, new SortByAge());
+                break;
+            case 2:
+                Arrays.sort(student, new SortByName());
+                break;
+            case 3:
+                Arrays.sort(student, new SortByRollNo());
+                break;
+            case 4:
+                Arrays.sort(student, new SortByAddress());
+        }
```
Modularize 
Input should be a different function 
Sorting should be in a different function
Output should be in a different function 
Single responsibility clause

ajayvenkat10 reviewed (changes requested):
Added comments

Commit:
Second Commit

Harshmanitripathi commented:
Sorry I forgot to Change the commit from second commit to third commit thats why it is showing like that.

New features: 
1. I added the validation and exception part .
2. Made the changes you told to do in previous commit except for delete operation because for delete operation I needed the index of the element which I wanted to delete.
3. Not able to properly validate the address because I don't know on what parameters we need to validate the address.
4. Done proper validation for courses and student count ,rollno,age,name.
5. Added personal class for exception handling with the name "BadDataEnteredException". Which just throws the type of error.

Next Work: 
I have started the third assignment because I need to complete that and one android assignment.

Commit:
Fourth Commit

Harshmanitripathi commented:
New Changes:
1. Removed the out file .
2. Made the choice for user to enter input as per the requirement whether he want to display(), delete(), sorting(), etc. In the Main class.
3. Performed the validation Operation and exception handling.

Commit:
Fifth Commit

Harshmanitripathi commented:
Removed the un-neccessary comments, from the **UserIO.java** file

ajayvenkat10 commented on src/StudentPackage/BadDataEnteredException.java:
```diff
@@ -0,0 +1,14 @@
+
+public class BadDataEnteredException extends Exception{
+
+    String s;
```
What is S ?
Don't use terminologies that we can't understand

Harshmanitripathi replied:
Declared variable with name "errorCameFromEnteringBadData"

ajayvenkat10 commented on src/StudentPackage/FileInputOutput.java:
```diff
@@ -0,0 +1,42 @@
+  public static void saveStudentDetailsInFile(Student[] students) throws FileNotFoundException {
+
+      try{
+          FileOutputStream fileOutPut = new FileOutputStream("obj.txt");
```
Accept filename from user as well

ajayvenkat10 commented on src/StudentPackage/FileInputOutput.java:
```diff
@@ -0,0 +1,42 @@
+          Student[] students1 = new Student[students.length];
+          FileInputStream fileInPut = new FileInputStream("obj.txt");
+          ObjectInputStream in = new ObjectInputStream(fileInPut);
+          students1 = (Student[]) in.readObject();
```
You'll need to read and store in a singleton class so that only one instance of it exists in a session

ajayvenkat10 commented on src/StudentPackage/FileInputOutput.java:
```diff
@@ -0,0 +1,42 @@
+          }
+
+      } catch (IOException e) {
+          throw new RuntimeException(e);
+      } catch (ClassNotFoundException e) {
+          throw new RuntimeException(e);
```
What is runtime exception ?
You need to create an exception for different types of exception 
Create custom exceptions like the BadDataEntered 
Also don't be very generic or have a broad range in a single type of exception 
Be specific if you can classify 
Either create different exceptions for different types 
Or create one exception class that has different exceptions you can call based on enum

Look what you are doing here, you are catching an IOException and displaying as runtime, how will you or a person working with you understand what sort of an exception it actually is ? You either print as the same exception or something more specific 
Same for ClassNotFound, pass the message to the exception that the given class is not found 

You can have an invalid input type custom exception that accepts an appropriate message 
For example:
1. If you accept only a character and if a user enters the whole string
2. You expect a string but the user enters an integer or so. You can throw out a custom message stating the given input is not of type string 
etc etc

Harshmanitripathi replied:
I have tried the custom exception class but it is showing Unhandled Exception, so I need to keep it

ajayvenkat10 replied:
That is for your throwable 
You are catching the right exception
Create a new custom throwable for you to throw after catching 
Throw this to the console with the appropriate message

ajayvenkat10 commented on src/StudentPackage/Main.java:
```diff
@@ -0,0 +1,40 @@
+    public static void main(String[] args) throws IOException {
+
+//        Giving Input from User
+        UserIO.TakingInput();
```
Already told you don't use present tense like taking, using etc.
Also told you that function names and variable names should start with small case 
Only class names should start with capital letters

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/Main.java:
```diff
@@ -0,0 +1,40 @@
+
+        Scanner sc = new Scanner(System.in);
+        while (true) {
+            System.out.print("Enter particular thing what you want sorting=1, display=2, deleting=3");
```
Each option on a different line

ajayvenkat10 replied:
Are these 3 the only options 
Provide all options

ajayvenkat10 commented on src/StudentPackage/Main.java:
```diff
@@ -0,0 +1,40 @@
+            String operationToPerform = sc.next();
+            switch (operationToPerform) {
+                case "1":
+                    UserIO.Sorting();
```
Again no "ing". No present tense
Just call sort ()

ajayvenkat10 commented on src/StudentPackage/Main.java:
```diff
@@ -0,0 +1,40 @@
+                    comeOut = true;
+                    break;
+            }
+            if (comeOut == true)
```
```comeOut == true```
You can just do if(comeOut) instead 
Not required to check the condition as the variable is a boolean

ajayvenkat10 commented on src/StudentPackage/Main.java:
```diff
@@ -0,0 +1,40 @@
+    public static void main(String[] args) throws IOException {
+
+//        Giving Input from User
+        UserIO.TakingInput();
```
Why is taking input separate and the rest of the operations separate ?
Have a menu 
Menu will have all possible operations 
Have a separate class whose method you'll call for input and output (For differentiating input and output with the rest) 
Just like how you have a class for File IO 
Basically you can have every method to be called inside UserIO and internally call whatever you want

ajayvenkat10 commented on src/StudentPackage/Student.java:
```diff
@@ -0,0 +1,101 @@
+public class Student implements Serializable{
+    public int rollNoOfStudent, ageOfStudent;
+    public String nameOfStudent, addressOfStudent;
+    //    boolean courseA, courseB, courseC, courseD, courseE, courseF;
```
Remove comments

ajayvenkat10 commented on src/StudentPackage/Student.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.Scanner;
+
+public class Student implements Serializable{
+    public int rollNoOfStudent, ageOfStudent;
+    public String nameOfStudent, addressOfStudent;
```
You have a constructor and a getter and setter 
You use getter and setter so that you can keep the members of the class as private 
So make them private, you are negating the point of using getters and setters by making them public

ajayvenkat10 commented on src/StudentPackage/Student.java:
```diff
@@ -0,0 +1,101 @@
+        System.out.println();
+    }
+
+//    public  void studentDeleted( int rollNoToBeDeleted, boolean checkForRollNo) {
+//
+//        while (checkForRollNo == false){
+//
+//
+//
+//                if (this.getRollNoOfStudent() == rollNoToBeDeleted){
+//                    checkForRollNo = true;
+//
+//                    for (int j=i;j<student.length-1;j++){
+//
+//                        student[j] = student[j+1];
+//                    }
+//                    student[student.length-1] = null;
+//                    break;
+//                }
+//
+//            if (checkForRollNo == false)
+//            {
+//
+//                System.out.println("Please enter the valid rollNo");
+//                Scanner sc = new Scanner(System.in);
+//                rollNoToBeDeleted = sc.nextInt();
+//            }
+//        }
+
+//    }
```
Remove commented out code

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+                continue;
+            }
+            break;
+        }
```
As discussed in the last meeting 
Validation should be in a different class

ajayvenkat10 replied:
Take input and pass to the respective method in the validation class 
That class will return if its a valid input or not. Throw an exception if its not at the Validation class itself 
Move forward with next steps if its a valid input

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+                    System.out.println(x.toString());
+                    continue;
+                }
+                break;
```
Validation should be part of a different class

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+                    rollNo1 = sc.next();
+                    boolean checkRollno = true;
+                    for (int j = 0; j < rollNo1.length(); j++) {
+                        if (!(rollNo1.charAt(j) >= '0' && rollNo1.charAt(j) <= '9')) {
```
There should be a better way to check if a character is a number or not

ajayvenkat10 replied:
Accept a number instead no

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+                }
+                break;
+            }
+
```
Move to validations class

ajayvenkat10 replied:
For everything you are checking the same as to whether its a number or not 
You are repeating code over and over again 
Redundant code 
Thats the reason you make it into a function in Validations class and call that here

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+                    System.out.println(x.toString());
+                    continue;
+                }
+                break;
```
Move to validations class

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+            }
+
+//            Validation for Address
+            while (true) {
+                try {
+                    System.out.print("Enter the address of student: ");
+                    addressOfStudent = sc.next();
+                    if (addressOfStudent.charAt(0) == ',')
+                        throw new BadDataEnteredException("Enter valid Address");
+                } catch (BadDataEnteredException x) {
+                    System.out.println(x.toString());
+                    continue;
+                }
+                break;
```
validations class

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+
+                try {
+                    while (true) {
+                        String inputFromUserAboutTheCourses = sc.next();
+                        if (((inputFromUserAboutTheCourses).toLowerCase()).equals("y") == true)
+                            break;
+                        if (Arrays.asList(setOfCourses).contains(inputFromUserAboutTheCourses.toUpperCase()) == true)
+                            System.out.println("This course you entered previously");
+                        else
+                            setOfCourses[k++] = inputFromUserAboutTheCourses.toUpperCase();
+
+                        String d = inputFromUserAboutTheCourses.toUpperCase();
+
+                        if (!(d.equals("A") || d.equals("B") || d.equals("C") || d.equals("D") || d.equals("E") || d.equals("F")))
+                            throw new BadDataEnteredException("Courses available are A,B,C,D,E,F");
+                        if (d.equals("Y"))
+                            if (!(k >= 4 && k <= 6))
+                                throw new BadDataEnteredException("Courses entered should be atleast 4 and less than 6");
+                    }
+                       }
+                      catch (BadDataEnteredException x){
+                           System.out.println(x.toString());
+                           continue;
+                       }
```
validations class

Harshmanitripathi replied:
Done

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+
+                        String d = inputFromUserAboutTheCourses.toUpperCase();
+
+                        if (!(d.equals("A") || d.equals("B") || d.equals("C") || d.equals("D") || d.equals("E") || d.equals("F")))
```
Rather check if the value is between Ascii value of A and F 
Also accommodate case insensitivity 
You should be forgiving here to accept even small cases 
So just convert the user's entered character to upper case and then apply the check 

Move the entire thing to validations class

Harshmanitripathi replied:
I have accepted smallcase letter in it & after accepting the value I converted those value to uppercase in variable "d".

ajayvenkat10 replied:
Rename 
What is 'd' ? 

And use the ascii value check

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+        FileInputOutput.saveStudentDetailsInFile(student);
+    }
+
+   static void Sorting(){
```
Call it displayWithSortOption instead

Harshmanitripathi replied:
done

ajayvenkat10 commented on src/StudentPackage/UserIO.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+   static void Sorting(){
+        System.out.println("enter the way you want to sort by age=1, name=2, rollno=3, address=4 etc.");
```
Dont do age = 1 
Display it like a menu 
And this goes to all the places where you display options 

Display like this : 

1. Age 
2. Name 
3. Roll no 
4. Address

Pointing out again 
This is not a college program anymore 
Its a project and you are building a product with features 
Would you be a happy user if your menu looked like how you are displaying currently 

Every step of the way think about the customer or end user checking out your product 
Communication should be clear and easily understandable 

Also when you code think of it as a project 
Now its only you coding, but think of it as a team project 
Someone else might code with you later, they will contribute 
It should be easily understandable and scalable 
Any function or any piece of code you write, first think if it can be generic. If it can then immediately move it to a class that fits its generic description if there already exists one or create one for it 
If its generic see if it should be easily accessible, then that method needs to be public 
Should it be accessible through an object or direct. If its direct then it needs to be a public static method.

Think every time :

Are you repeating code ? 
Are you giving multiple responsibilities to a single function ? 
Can you modularize more ? 
Can this piece of code be generic ? 
Are you using the right access specifiers for every variable and function you write ?
Is it easy for others to understand and reuse ?
Is it easy for the customer to understand ?
Are you validating correctly ?
Are you optimizing correctly ?
Are you throwing the right exceptions which can be easily utilised by someone else in the future ?
Are your exception messages easy to understand for you, other devs and your customer so that they exactly understand what is the problem and rectify their mistake ?

Harshmanitripathi replied:
Thank you explaning things in details , I will try to keep all this things in mind.

ajayvenkat10 reviewed (changes requested):
Added comments

Commit:
Sixth Commit

Harshmanitripathi commented:
Changes made:
1. Changed s name to -> errorCameFromEnteringBadData.
2. added singleton class to store data
3. added validation class
4. changed the name to function to small letter

Things not sorted:
1.Not able to display the data from file .
